lib/actions/os.js

```diff
@@ -0,0 +1,353 @@
+	$ balena os download raspberrypi3 -o ../foo/bar/raspberry-pi.img --version default
+	$ balena os download raspberrypi3 -o ../foo/bar/raspberry-pi.img --version menu\
+`,
+	permission: 'user',
+	options: [
+		{
+			signature: 'output',
+			description: 'output path',
+			parameter: 'output',
+			alias: 'o',
+			required: 'You have to specify the output location',
+		},
+		commandOptions.osVersionOrSemver,
+	],
+	action(params, options) {
+		const Promise = require('bluebird');
+		const unzip = require('node-unzip-2');
+		const fs = require('fs');
+		const rindle = require('rindle');
+		const manager = require('balena-image-manager');
+
+		console.info(`Getting device operating system for ${params.type}`);
+
+		let displayVersion = '';
+		return Promise.try(function() {
+			if (!options.version) {
+				console.warn(`OS version is not specified, using the default version: \
+the newest stable (non-pre-release) version if available, \
+or the newest version otherwise (if all existing \
+versions for the given device type are pre-release).`);
+				return 'default';
+			}
+			return resolveVersion(params.type, options.version);
+		})
+			.then(function(version) {
+				if (version !== 'default') {
+					displayVersion = ` ${version}`;
+				}
+				return manager.get(params.type, version);
+			})
+			.then(function(stream) {
+				const visuals = getVisuals();
+				const bar = new visuals.Progress(
+					`Downloading Device OS${displayVersion}`,
+				);
+				const spinner = new visuals.Spinner(
+					`Downloading Device OS${displayVersion} (size unknown)`,
+				);
+
+				stream.on('progress', function(state) {
+					if (state != null) {
+						return bar.update(state);
+					} else {
+						return spinner.start();
+					}
+				});
+
+				stream.on('end', () => {
+					spinner.stop();
+				});
+
+				// We completely rely on the `mime` custom property
+				// to make this decision.
+				// The actual stream should be checked instead.
+				let output;
+				if (stream.mime === 'application/zip') {
+					output = unzip.Extract({ path: options.output });
+				} else {
+					output = fs.createWriteStream(options.output);
+				}
+
+				return rindle.wait(stream.pipe(output)).return(options.output);
+			})
+			.tap(() => {
+				console.info('The image was downloaded successfully');
+			});
+	},
+};
+
+const buildConfigForDeviceType = function(deviceType, advanced) {
+	if (advanced == null) {
+		advanced = false;
+	}
+	const form = require('resin-cli-form');
+	const helpers = require('../utils/helpers');
+
+	let override;
+	const questions = deviceType.options;
+	if (!advanced) {
+		const advancedGroup = _.find(questions, {
+			name: 'advanced',
+			isGroup: true,
+		});
+
+		if (advancedGroup != null) {
+			override = helpers.getGroupDefaults(advancedGroup);
+		}
+	}
+
+	return form.run(questions, { override });
+};
+
+const $buildConfig = function(image, deviceTypeSlug, advanced) {
```

Comment:
I had to rename this to `$buildConfig` to avoid clashing with the exported `buildConfig`